openmessaging-format.md

```diff
@@ -0,0 +1,130 @@
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][RFC2119].
+
+## 2. Attributes
+
+This section defines how CloudEvents attributes are mapped to OpenMessaging. This
+specification does not explicitly map each attribute, but provides a generic
+mapping model that applies to all current and future CloudEvents attributes.
+
+### 2.1. Base Type System
+
+The core [CloudEvents specification][CE] defines a minimal abstract type
+system, which this mapping leans on.
+
+### 2.2. Type System Mapping
+
+The CloudEvents type system is mapped to JSON types as follows:
+
+| CloudEvents | JSON
+|--------------|-------------------------------------------------------------
+| String       | String
+| Binary       | Binary
+| URI          | String
+| Timestamp    | String
+| Map          | Map
+| Object       | see2.3
+
+### 2.3. Mapping Object-typed Attributes
+
+`Object`-typed CloudEvents values can either hold a `String`, or a `Binary`
+value, or a `Map`. `Map` entry values are also `Object` typed. OpenMessaging's type
+system natively represents dynamic typing in its [type system
+encoding][type-system-encoding], and therefore immediately allows for the required
+variant type representation.
+
+### 2.4. Examples
+
+The following table shows exemplary mappings:
+
+| CloudEvents       | Type     | Exemplary OpenMessaging Value
+|--------------------|----------|-------------------------------
+| eventType          | String   | "openMessaging"
+| eventTypeVersion   | String   | "1.0"
+| cloudEventsVersion | String   | "0.1"
+| source             | URI      | "openMessaging.io.event"
+| eventID            | String   | "1234-1234-1234"
+| eventTime          | Timestamp| "2018-04-05T17:31:00Z"
+| contentType        | String   | "application/json"
+| extensions         | Map      | { "extA" : "vA", "extB", "vB" }
+| data               | String   | "test openMessaging"
+| data               | Binary   | "156,134,111"
+| data               | Map      | { "objA" : "vA", "objB", "vB" }
+
+## 3. Envelope
+Each CloudEvents event can be wholly represented as a OpenMessaging object.
+
+Such a representation uses the *userHeader* in the OpenMessaging,
+If the property "cloudEvent" existed, it represented it's a cloud event message.
+All REQUIRED and all not omitted OPTIONAL attributes in the given event MUST
+become members of the OpenMessaging object, with the respective OpenMessaging object member name
+matching the attribute name, and the member's type and value being mapped using
+the [type system mapping](#22-type-system-mapping).
+
+### 3.1. Special Handling of the "data" Attribute
+
+The mapping of the `Object`-typed `data` attribute follows the rules laid out
+in [Section 2.3.](#23-mapping-object-typed-attributes), with one additional
+rule:
+
+If an implementation determines that the type of the `data` attribute is
+`Binary` or `String`, it MUST inspect the `contentType` attribute to determine
```

Comment:
I'm not sure this MUST is required since our specs don't really get into the processing model. Perhaps instead of defining a new rule we turn this more into "guidance":

The mapping of the `Object`-typed `data` attribute follows the rules laid out in Section 2.3. Receiving
implementations can use the `contentType` attribute to determine the format of value within
the `data` attribute.

Although, TBH, I'm not sure this is needed since that's the point of the contentType attribute so I'm not sure its necessary to repeat it here.

Reply:
Actually, can we just point to the json-format.md doc instead of repeating this entire section?

Reply:
Next, we will refine these descriptions, to make the definition of data type clear and understandable.